File: practices/interview-bit/stacks-and-queues/rain-water-trapped.py
# Solution 1: dynamic programming
# O(N) in time
# O(N) in space
class Solution:
    # @param arr: tuple of integers
    # @return an integer
    def trap(self, arr):
        n = len(arr)

        maxLeft = [0] * n
        maxLeft[0] = arr[0]
        for i in range(1, n):
            maxLeft[i] = max(maxLeft[i - 1], arr[i])

        maxRight = [0] * n
        maxRight[n - 1] = arr[n - 1]
        for i in range(n-2,-1,-1):
            maxRight[i] = max(maxRight[i + 1], arr[i])

        water = 0
        for i in range(0,n):
            water += min(maxLeft[i], maxRight[i]) - arr[i]
        return water

# Solution 1: dynamic programming
# O(N) in time
# O(N) in space
class Solution:
    def trap(self, arr):
        n = len(arr)

        left = 0
        right = n - 1

        maxLeft = 0
        maxRight = 0

        water = 0
        while left <= right:
            if arr[left] <= arr[right]:
                maxLeft = max(maxLeft, arr[left])
                water += maxLeft - arr[left]
                left += 1
            else:
                maxRight = max(maxRight, arr[right])
                water += maxRight - arr[right]
                right -= 1
        return water 

File: practices/interview-bit/stacks-and-queues/rain-water-trapped.cpp
// Solution 1: dynamic programming
// O(N) in time
// O(N) in space
int Solution::trap(const vector<int> &A) {
    int n = (int)A.size();
    if (n <= 2) return 0;

    vector<int> maxLeft(n);
    maxLeft[0] = INT_MIN;
    for(int i = 1; i < n; i++)
        maxLeft[i] = max(A[i - 1], maxLeft[i - 1]);

    vector<int> maxRight(n);
    maxRight[n - 1] = INT_MIN;
    for(int i = n - 2; i >= 0; i--)
        maxRight[i] = max(A[i + 1], maxRight[i + 1]);

    int ans = 0;
    for(int i = 0; i < n; i++) {
        int minBoth = min(maxLeft[i], maxRight[i]);
        if (minBoth > A[i]) ans += (minBoth - A[i]);
    }
    return ans;
}

// Solution 2: two pointers
// O(N) in time
// O(1) in space
int Solution::trap(const vector<int> &A) {
    int left = 0, right = (int)A.size() - 1;
    int maxLeft = 0, maxRight = 0;
    int ans = 0;
    while(left <= right) {
        if (A[left] <= A[right]) {
            if (A[left] > maxLeft) maxLeft = A[left];
            else ans += (maxLeft - A[left]);  // Because we move the lower-value pointer,
                                              // and the 'max...' is only updated after that,
                                              // maxLeft is always < A[right]
            left++;
        } else {
            if (A[right] > maxRight) maxRight = A[right];
            else ans += (maxRight - A[right]);
            right--;
        }
    }

    return ans;
}
